File: backend/models/tournament.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type TournamentStatus string

const (
	TournamentStatusLobby    TournamentStatus = "lobby"
	TournamentStatusActive   TournamentStatus = "active"
	TournamentStatusFinished TournamentStatus = "finished"
)

type TournamentFormat string

const (
	TournamentFormatSingleElimination TournamentFormat = "single_elimination"
	TournamentFormatDoubleElimination TournamentFormat = "double_elimination"
	TournamentFormatGroupKnockout     TournamentFormat = "group_knockout"
)

type Tournament struct {
	gorm.Model
	Name            string           `json:"name"`
	Slug            string           `json:"slug" gorm:"uniqueIndex;size:24"`
	HostUserID      uint             `json:"hostUserId" gorm:"index"`
	HostUser        User             `json:"hostUser"`
	Status          TournamentStatus `json:"status" gorm:"type:varchar(24);default:'lobby';index"`
	Format          TournamentFormat `json:"format" gorm:"type:varchar(32);default:'single_elimination'"`
	GrandFinalReset *bool            `json:"grandFinalReset,omitempty"`
	GroupSize       *int             `json:"groupSize,omitempty"`
	AdvancePerGroup *int             `json:"advancePerGroup,omitempty"`
	MaxPlayers      int              `json:"maxPlayers" gorm:"default:16"`
	IsLocked        bool             `json:"isLocked" gorm:"default:false"`
	InviteTokenHash string           `json:"-" gorm:"type:char(64);index"`
	StartedAt       *time.Time       `json:"startedAt,omitempty"`
	EndedAt         *time.Time       `json:"endedAt,omitempty"`
}

func (Tournament) TableName() string {
	return "tournaments"
}

File: frontend/src/types/tournament.ts
export type TournamentStatus = 'lobby' | 'active' | 'finished'
export type TournamentFormat = 'single_elimination' | 'double_elimination' | 'group_knockout'

export type TournamentSettings = {
  grandFinalReset?: boolean | null
  groupSize?: number | null
  advancePerGroup?: number | null
}

export type TournamentSummary = {
  id: number
  name: string
  slug: string
  hostUserId: number
  status: TournamentStatus
  format: TournamentFormat
  settings?: TournamentSettings
  maxPlayers: number
  isLocked: boolean
  startedAt?: string | null
  endedAt?: string | null
  createdAt?: string
  updatedAt?: string
}

export type TournamentParticipant = {
  id: number
  tournamentId: number
  userId?: number | null
  displayName: string
  avatarUrl: string
  seed?: number | null
  joinedAt: string
  eliminatedAt?: string | null
  isHost: boolean
}

export type TournamentMatch = {
  ID: number
  tournamentId: number
  round: number
  slot: number
  stageType?: 'knockout' | 'group' | 'grand_final'
  bracketType?: 'main' | 'winners' | 'losers'
  groupNumber?: number | null
  stageRound?: number | null
  playerAParticipantId?: number | null
  playerBParticipantId?: number | null
  winnerParticipantId?: number | null
  status: 'pending' | 'playing' | 'completed'
  liveMatchId?: string | null
  startedAt?: string | null
  finishedAt?: string | null
}

export type TournamentLeaderboardEntry = {
  participantId: number
  displayName: string
  avatarUrl: string
  seed?: number
  joinedAt: string
  wins: number
  losses: number
  points: number
}

export type CreateTournamentRequest = {
  name: string
  maxPlayers: number
  format?: TournamentFormat
  settings?: TournamentSettings
}

export type CreateTournamentResponse = {
  tournament: TournamentSummary
  inviteToken: string
  inviteLink: string
}

export type JoinTournamentRequest = {
  inviteToken: string
  displayName?: string
  avatarUrl?: string
}

export type JoinTournamentResponse = {
  tournament: TournamentSummary
  participant: TournamentParticipant
  guestSessionToken?: string
  tournamentAccessBy: string
}

export type TournamentBySlugResponse = {
  tournament: TournamentSummary
  participant?: TournamentParticipant
  participants?: TournamentParticipant[]
  matches?: TournamentMatch[]
  requiresJoin: boolean
  inviteTokenParam?: string
}

export type TournamentDetailsResponse = {
  tournament: TournamentSummary
  participant: TournamentParticipant
  participants: TournamentParticipant[]
  matches: TournamentMatch[]
}
